multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ApplicationZipBuilder.java

```diff
@@ -1,113 +1,180 @@
+                saveAsZip(fileOutputStream, applicationArchiveContext, zipEntry, zipArchiveInputStream);
             } else {
-                saveToFile(fileOutputStream, applicationArchiveContext, zipEntry);
+                saveToFile(fileOutputStream, applicationArchiveContext, zipEntry, zipArchiveInputStream);
             }
         }
     }
 
-    private void saveAsZip(OutputStream fileOutputStream, ApplicationArchiveContext applicationArchiveContext, ZipEntry zipEntry)
+    private void saveAsZip(OutputStream fileOutputStream, ApplicationArchiveContext applicationArchiveContext, ZipEntry zipEntry,
+                           ZipArchiveInputStream zipArchiveInputStream)
         throws IOException {
         try (ZipOutputStream zipOutputStream = new ZipOutputStream(fileOutputStream)) {
             String moduleFileName = applicationArchiveContext.getModuleFileName();
             do {
                 if (!isAlreadyUploaded(zipEntry.getName(), applicationArchiveContext) && !zipEntry.isDirectory()) {
                     zipOutputStream.putNextEntry(createNewZipEntry(zipEntry.getName(), moduleFileName));
-                    copy(applicationArchiveContext.getZipInputStream(), zipOutputStream, applicationArchiveContext);
+                    copy(zipArchiveInputStream, zipOutputStream, applicationArchiveContext);
                     zipOutputStream.closeEntry();
-
                 }
-            } while ((zipEntry = applicationArchiveReader.getNextEntryByName(moduleFileName, applicationArchiveContext)) != null);
+            } while ((zipEntry = applicationArchiveIterator.getNextEntryByName(moduleFileName, zipArchiveInputStream)) != null);
         }
     }
 
-    private ZipEntry createNewZipEntry(String zipEntryName, String moduleFileName) {
-        return new UtcAdjustedZipEntry(FileUtils.getRelativePath(moduleFileName, zipEntryName));
-    }
-
-    private void saveToFile(OutputStream fileOutputStream, ApplicationArchiveContext applicationArchiveContext, ZipEntry zipEntry)
-        throws IOException {
-        String moduleFileName = applicationArchiveContext.getModuleFileName();
-        do {
-            if (!isAlreadyUploaded(zipEntry.getName(), applicationArchiveContext)) {
-                copy(applicationArchiveContext.getZipInputStream(), fileOutputStream, applicationArchiveContext);
-            }
-        } while ((zipEntry = applicationArchiveReader.getNextEntryByName(moduleFileName, applicationArchiveContext)) != null);
-    }
-
     private boolean isAlreadyUploaded(String zipEntryName, ApplicationArchiveContext applicationArchiveContext) {
         return applicationArchiveContext.getAlreadyUploadedFiles()
                                         .contains(zipEntryName);
     }
 
+    private ZipEntry createNewZipEntry(String zipEntryName, String moduleFileName) {
+        return new UtcAdjustedZipEntry(FileUtils.getRelativePath(moduleFileName, zipEntryName));
```

Comment:
Why do we need UtcAdjustedZipEntry because of the creation time or?

Reply:
we set the modification time of the entry
@theghost5800 could you provide input :D if you remember